staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go

```diff
@@ -0,0 +1,167 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-openapi/spec"
+	"github.com/golang/glog"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kube-aggregator/pkg/apis/apiregistration"
+)
+
+const (
+	successfulUpdateDelay   = time.Minute
+	failedUpdateMaxExpDelay = time.Hour
+)
+
+// OpenAPIAggregationManager is the interface between this controller and OpenAPI Aggregator service.
+type OpenAPIAggregationManager interface {
+	AddUpdateApiService(handler http.Handler, apiService *apiregistration.APIService) error
+	UpdateApiServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error
+	RemoveApiServiceSpec(apiServiceName string) error
+	GetApiServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool)
+}
+
+// OpenAPIAggregationController periodically check for changes in OpenAPI specs of APIServices and update/remove
+// them if necessary.
+type OpenAPIAggregationController struct {
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+	downloader                *openAPIDownloader
+
+	// To allow injection for testing.
+	syncHandler func(key string) error
+}
+
+func NewOpenAPIAggregationController(downloader *openAPIDownloader, openAPIAggregationManager OpenAPIAggregationManager) *OpenAPIAggregationController {
+	c := &OpenAPIAggregationController{
+		openAPIAggregationManager: openAPIAggregationManager,
+		queue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.NewItemExponentialFailureRateLimiter(successfulUpdateDelay, failedUpdateMaxExpDelay), "APIServiceOpenAPIAggregationControllerQueue1"),
+		downloader:  downloader,
+	}
+
+	c.syncHandler = c.sync
+
+	return c
+}
+
+func (c *OpenAPIAggregationController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
+	defer glog.Infof("Shutting down OpenAPIAggregationController")
+
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *OpenAPIAggregationController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	defer c.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		return true
```

Comment:
Should we forget the key to reset rate limiting for that key? Most controllers do so, or have an explicit comment here if they don't.